### ISSUE TITLE
DataParallel replicates network object for device 0

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#11476]
This makes some hooks just not work, e.g., if your hook sets some attribute on the `module` input, they are lost. E.g., spectral norm. This makes DataParallel + spectral norm not converging.

Also DataParallel + spectral norm doesn't work in eval mode because we don't properly broadcast buffers that require grad. See the following repro:

```
import torch
import torch.nn as nn
from torch.nn.utils import spectral_norm

net = nn.Conv2d(3, 3, 3).cuda()
net = spectral_norm(net)
net = nn.DataParallel(net)

inp = torch.randn(2, 3, 4, 4).cuda()

net(inp)

net.eval()
net(inp)
```

I'll fix both.

**Comment by YaoshengFu** [comment]
I have encountered the same issue when I tried to run multi-gpu training with DataParallel + spectral norm but it works fine for single gpu. Is there any progress on fixing this?

**Comment by SsnL** [comment]
I'll be submitting a patch soon @YaoshengFu

**Comment by YaoshengFu** [comment]
That would be great, thanks!
